File: 4.Authentication and authorization for backend engineers/code/python/05_json_web_tokens_jwt_4.py
# Verify with a PUBLIC key, pinning RS256 (blocks alg:none & key confusion)
def verify_rs256(token: str, public_key) -> dict:
    return jwt.decode(token, public_key, algorithms=["RS256"])  # whitelist

# Rotate a refresh token; detect reuse of an already-spent one.
def refresh(rdb, presented: str) -> str:
    family = rdb.get(f"rt:{presented}")
    if family is None:
        # not a live token ,  was it previously spent? -> theft
        fam = rdb.get(f"spent:{presented}")
        if fam is not None:
            rdb.delete(f"family:{fam.decode()}")    # revoke the whole family
            raise PermissionError("refresh reuse detected")
        raise PermissionError("invalid refresh token")
    rdb.delete(f"rt:{presented}")                   # consume the live token
    rdb.setex(f"spent:{presented}", 14 * 24 * 3600, family)  # remember it
    new_refresh = secrets.token_hex(32)
    rdb.setex(f"rt:{new_refresh}", 14 * 24 * 3600, family)
    return new_refresh

File: 4.Authentication and authorization for backend engineers/code/go/05_json_web_tokens_jwt_4.go
// Verify with a PUBLIC key, pinning RS256 (blocks alg:none & key confusion)
func VerifyRS256(tokenStr string, pub *rsa.PublicKey) (jwt.MapClaims, error) {
    tok, err := jwt.Parse(tokenStr, func(t *jwt.Token) (any, error) {
        if _, ok := t.Method.(*jwt.SigningMethodRSA); !ok { // pin family
            return nil, jwt.ErrSignatureInvalid
        }
        return pub, nil
    }, jwt.WithValidMethods([]string{"RS256"}))
    if err != nil || !tok.Valid {
        return nil, err
    }
    return tok.Claims.(jwt.MapClaims), nil
}

// Rotate a refresh token; detect reuse of an already-spent one.
func Refresh(rdb *redis.Client, presented string) (string, error) {
    family, err := rdb.Get(ctx, "rt:"+presented).Result()
    if err != nil {
        // not a live token ,  was it a previously-spent one? -> theft
        if fam, e := rdb.Get(ctx, "spent:"+presented).Result(); e == nil {
            rdb.Del(ctx, "family:"+fam)            // revoke the whole family
            return "", errors.New("refresh reuse detected")
        }
        return "", errors.New("invalid refresh token")
    }
    rdb.Del(ctx, "rt:"+presented)                  // consume the live token
    rdb.Set(ctx, "spent:"+presented, family, 14*24*time.Hour) // remember it
    newRefresh := newSessionID()
    rdb.Set(ctx, "rt:"+newRefresh, family, 14*24*time.Hour)
    return newRefresh, nil
}
